MINOR: Improve Log layer segment iteration logic and few other areas

kowshik opened the pull request:
1. In `Log.collectAbortedTransactions()` I've restored a previously used logic, such that it would handle the case where the starting segment could be null. This was the case [previously](https://github.com/apache/kafka/blob/8205051e90e3ea16165f8dc1f5c81af744bb1b9a/core/src/main/scala/kafka/log/Log.scala#L1658-L1661), but the PR https://github.com/apache/kafka/pull/10401 accidentally [changed the behavior](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/Log.scala#L1240-L1243) causing the code to assume that the starting segment won't be null.
2. In `Log.rebuildProducerState()` I've removed usage of the `allSegments` local variable. The logic looks a bit simpler after I removed it.
3. I've introduced a new `LogSegments.higherSegments()` API. This is now used to make the logic a bit more readable in `Log. collectAbortedTransactions()` and `Log.deletableSegments()` APIs.
4. I've removed the unnecessary use of `java.lang.Long` in `LogSegments` class' segments map definition.
5. I've converted a few `LogSegments` API from public to private, as they need not be public.

**Tests:**
Relying on existing unit tests.

kowshik committed:
MINOR: Improve Log layer segment iteration logic and few other areas

kowshik committed:
Minor improvement

kowshik commented on core/src/main/scala/kafka/log/LogSegments.scala:
```diff
@@ -210,11 +209,23 @@ class LogSegments(topicPartition: TopicPartition) {
+   * Returns an iterable to log segments ordered from lowest base offset to highest.
+   * Each segment in the returned iterable has a base offset strictly greater than the provided baseOffset.
+   */
+  def higherSegments(baseOffset: Long): Iterable[LogSegment] = {
```
@ijuma If you think this API is a good fit to the requirement, I can add unit tests for it.

kowshik commented:
cc @junrao

kowshik commented on core/src/main/scala/kafka/log/LogSegments.scala:
```diff
@@ -17,7 +17,6 @@
 package kafka.log
 
 import java.io.File
-import java.lang.{Long => JLong}
```
cc @ccding

ccding commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1204,13 +1202,12 @@ class Log(@volatile private var _dir: File,
-        var done = segmentEntryOpt.isEmpty
+        var done = segmentOpt.isEmpty
         var fetchDataInfo: FetchDataInfo = null
+        val segmentsIterator = segmentOpt.map(segment => segments.higherSegments(segment.baseOffset))
```
Given we can get baseOffset from segment, can we get rid of `floorEntry` overall? or make `floorEntry` private?

kowshik replied:
Done, made it private. Please see 80c40171302546be2dcf2444167d7f375f0b820d.

ccding commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1219,10 +1216,10 @@ class Log(@volatile private var _dir: File,
+          } else segmentOpt = segmentsIterator.nextOption()
 
-          done = fetchDataInfo != null || segmentEntryOpt.isEmpty
+          done = fetchDataInfo != null || segmentOpt.isEmpty
```
can we simply use `while (fetchDataInfo == null && segmentOpt.isDefined)` rather than `while (!done)`?

kowshik replied:
Done. Please see 80c40171302546be2dcf2444167d7f375f0b820d.

ijuma commented on core/src/main/scala/kafka/log/LogSegments.scala:
```diff
@@ -185,7 +184,7 @@ class LogSegments(topicPartition: TopicPartition) {
    */
   @threadsafe
-  def higherEntry(offset: Long): Option[Map.Entry[JLong, LogSegment]] = Option(segments.higherEntry(offset))
+  def higherEntry(offset: Long): Option[Map.Entry[Long, LogSegment]] = Option(segments.higherEntry(offset))
```
Why do we use `Map.Entry` instead of tuple here? Is it to avoid the allocation?

kowshik replied:
That's one reason. The other is that `Map.Entry` provided a better interface for attribute lookup via `Map.Entry.getKey()` and `Map.Entry.getValue()`, as compared to `._1` and `._2` in a tuple.

kowshik committed:
Address comments

ccding commented:
LGTM

Thanks @kowshik

kowshik committed:
Fix inconsistent Option/null check in Log.collectAbortedTransactions

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1217,10 +1213,8 @@ class Log(@volatile private var _dir: File,
-
-          done = fetchDataInfo != null || segmentEntryOpt.isEmpty
+              fetchDataInfo = addAbortedTransactions(startOffset, segment, fetchDataInfo)
+          } else segmentOpt = segmentsIterator.nextOption()
```
The old logic supports skipping forward multiple segments to find the right data. The new logic seems to only support skipping forward once. It would be useful to preserve the original semantic.

kowshik replied:
That is an excellent catch. Sorry I missed this. Done, changed it now.

junrao commented on core/src/main/scala/kafka/log/LogSegments.scala:
```diff
@@ -210,11 +209,23 @@ class LogSegments(topicPartition: TopicPartition) {
+    val view =
+      Option(segments.higherKey(baseOffset)).map {
+        higherOffset => segments.tailMap(higherOffset, true)
+      }.getOrElse(new ConcurrentSkipListMap[Long, LogSegment]())
```
Could we return a constant empty map?

kowshik replied:
Done.

junrao reviewed (commented):
@kowshik : Thanks for the PR. A couple of comments below.

kowshik committed:
Address comments from Jun

kowshik commented:
Thanks for the review @junrao! I have addressed your comments in 4b836034415c3d5f6b84384ef9be1e75b66edc4b.

junrao reviewed (commented):
@kowshik : Thanks for the updated PR. There seems to be a compilation error in java 8.

kowshik committed:
Fix build in scala 2.12 and add unit tests

kowshik commented:
@junrao Thanks for the review and for catching this issue! I've fixed the build now. The issue was that the code in `Log.scala` used the `Iterator.nextOption()`[API](https://www.scala-lang.org/api/2.13.0/scala/collection/Iterator.html#nextOption():Option[A]) that was introduced only in scala v2.13, so the build failed for scala v2.12.

I have now also added unit tests for the newly introduced `LogSegments.higherSegments()` API.

The PR is ready for review again.

junrao reviewed (approved):
@kowshik : Thanks for the updated PR. LGTM
